File: src/main/java/io/github/dylmeadows/eontimer/controller/timer/Gen3TimerController.java
package io.github.dylmeadows.eontimer.controller.timer;

import io.github.dylmeadows.eontimer.model.timer.Gen3TimerConstants;
import io.github.dylmeadows.eontimer.model.timer.Gen3TimerMode;
import io.github.dylmeadows.eontimer.model.timer.Gen3TimerModel;
import io.github.dylmeadows.eontimer.util.Spinners;
import io.github.dylmeadows.common.javafx.util.ChoiceConverter;
import io.github.dylmeadows.common.javafx.util.Nodes;
import javafx.beans.binding.BooleanBinding;
import javafx.collections.FXCollections;
import javafx.fxml.FXML;
import javafx.scene.control.ChoiceBox;
import javafx.scene.control.Spinner;
import javafx.scene.control.SpinnerValueFactory;
import javafx.scene.layout.VBox;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

@Component
@RequiredArgsConstructor(onConstructor_ = @Autowired)
public class Gen3TimerController {

    private final Gen3TimerModel model;

    @FXML
    private ChoiceBox<Gen3TimerMode> modeField;
    @FXML
    private Spinner<Integer> calibrationField;
    @FXML
    private Spinner<Integer> preTimerField;
    @FXML
    private Spinner<Integer> targetFrameField;
    @FXML
    private Spinner<Integer> frameHitField;

    @FXML
    private VBox calibrationFieldSet;
    @FXML
    private VBox frameHitFieldSet;

    public void initialize() {
        // Mode
        modeField.setItems(FXCollections.observableArrayList(Gen3TimerMode.values()));
        modeField.setConverter(ChoiceConverter.forChoice(Gen3TimerMode.class));
        modeField.valueProperty().bindBidirectional(model.modeProperty());
        // Calibration
        SpinnerValueFactory<Integer> calibrationValueFactory = Spinners.createValueFactory(
            Integer.MIN_VALUE, Integer.MAX_VALUE, Gen3TimerConstants.DEFAULT_CALIBRATION);
        calibrationValueFactory.valueProperty().bindBidirectional(model.calibrationProperty().asObject());
        calibrationField.setValueFactory(calibrationValueFactory);
        // Pre-Timer
        SpinnerValueFactory<Integer> preTimerValueFactory = Spinners.createValueFactory(
            0, Integer.MAX_VALUE, Gen3TimerConstants.DEFAULT_PRE_TIMER);
        preTimerValueFactory.valueProperty().bindBidirectional(model.preTimerProperty().asObject());
        preTimerField.setValueFactory(preTimerValueFactory);
        // Target Frame
        SpinnerValueFactory<Integer> targetFrameValueFactory = Spinners.createValueFactory(
            0, Integer.MAX_VALUE, Gen3TimerConstants.DEFAULT_TARGET_FRAME);
        targetFrameValueFactory.valueProperty().bindBidirectional(model.targetFrameProperty().asObject());
        targetFrameField.setValueFactory(targetFrameValueFactory);
        // Frame Hit
        SpinnerValueFactory<Integer> frameHitValueFactory = Spinners.createValueFactory(0, Integer.MAX_VALUE);
        frameHitValueFactory.valueProperty().bindBidirectional(model.frameHitProperty().asObject());
        frameHitField.setValueFactory(frameHitValueFactory);

        final BooleanBinding isStandardMode = modeField.valueProperty().isNotEqualTo(Gen3TimerMode.STANDARD);
        Nodes.hideAndResizeParentIf(calibrationFieldSet, isStandardMode);
        Nodes.hideAndResizeParentIf(frameHitFieldSet, isStandardMode);
    }
}

File: src/main/kotlin/io/github/dylmeadows/eontimer/controller/timer/Gen3TimerController.kt
package io.github.dylmeadows.eontimer.controller.timer

import io.github.dylmeadows.common.javafx.util.ChoiceConverter
import io.github.dylmeadows.common.javafx.util.Nodes
import io.github.dylmeadows.eontimer.model.timer.Gen3TimerConstants
import io.github.dylmeadows.eontimer.model.timer.Gen3TimerMode
import io.github.dylmeadows.eontimer.model.timer.Gen3TimerModel
import io.github.dylmeadows.eontimer.util.createValueFactory
import javafx.collections.FXCollections
import javafx.fxml.FXML
import javafx.scene.control.ChoiceBox
import javafx.scene.control.Spinner
import javafx.scene.layout.VBox
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component

@Component
class Gen3TimerController @Autowired constructor(
    private val model: Gen3TimerModel) {

    @FXML
    private lateinit var modeField: ChoiceBox<Gen3TimerMode>
    @FXML
    private lateinit var calibrationField: Spinner<Int>
    @FXML
    private lateinit var preTimerField: Spinner<Int>
    @FXML
    private lateinit var targetFrameField: Spinner<Int>
    @FXML
    private lateinit var frameHitField: Spinner<Int>

    @FXML
    private lateinit var calibrationFieldSet: VBox
    @FXML
    private lateinit var frameHitFieldSet: VBox

    fun initialize() {
        modeField.items = FXCollections.observableArrayList(*Gen3TimerMode.values())
        modeField.converter = ChoiceConverter.forChoice(Gen3TimerMode::class.java)
        modeField.valueProperty().bindBidirectional(model.modeProperty)

        calibrationField.createValueFactory(Int.MIN_VALUE, Int.MAX_VALUE, Gen3TimerConstants.DEFAULT_CALIBRATION)
            .valueProperty().bindBidirectional(model.calibrationProperty.asObject())
        preTimerField.createValueFactory(0, Int.MAX_VALUE, Gen3TimerConstants.DEFAULT_PRE_TIMER)
            .valueProperty().bindBidirectional(model.preTimerProperty.asObject())
        targetFrameField.createValueFactory(0, Int.MAX_VALUE, Gen3TimerConstants.DEFAULT_TARGET_FRAME)
            .valueProperty().bindBidirectional(model.targetFrameProperty.asObject())
        frameHitField.createValueFactory(0, Int.MAX_VALUE)
            .valueProperty().bindBidirectional(model.frameHitProperty.asObject())

        // set conditional field visibility
        val isStandardMode = modeField.valueProperty().isNotEqualTo(Gen3TimerMode.STANDARD)
        Nodes.hideAndResizeParentIf(calibrationFieldSet, isStandardMode)
        Nodes.hideAndResizeParentIf(frameHitFieldSet, isStandardMode)
    }
}